controllers/volumereplicationgroup_controller.go

```diff
@@ -1029,74 +1146,313 @@ func (v *VRGInstance) createVR(vrNamespacedName types.NamespacedName, state volr
+			// that saw some error, then entire VRG
+			// should mark its condition as error. So
+			// set vrgPogressing to false.
+			vrgProgressing = false
+
+			break
+		}
+	}
+
+	if vrgReady {
+		v.log.Info("Marking VRG available with replicating reason")
+
+		msg := "PVCs in the VolumeReplicationGroup group are replicating"
+		setVRGReplicatingCondition(&v.instance.Status.Conditions, v.instance.Generation, msg)
+
+		return
+	}
+
+	if vrgProgressing {
+		v.log.Info("Marking VRG not available with progressing reason")
+
+		msg := "VolumeReplicationGroup is progressing"
+		setVRGProgressingCondition(&v.instance.Status.Conditions, v.instance.Generation, msg)
+
+		return
 	}
 
+	// None of the VRG Ready and VRG Progressing conditions are met.
+	// Set Error condition for VRG.
+	v.log.Info("Marking VRG not available with error. All PVCs are not ready")
+
+	msg := "All PVCs of the VolumeReplicationGroup are not available"
+	setVRGErrorCondition(&v.instance.Status.Conditions, v.instance.Generation, msg)
+}
+
+func (v *VRGInstance) checkVRStatus(volRep *volrep.VolumeReplication) error {
 	switch {
-	case v.instance.Spec.ReplicationState == volrep.Primary:
+	case v.instance.Spec.ReplicationState == ramendrv1alpha1.Primary:
 		return v.validateVRStatusAsPrimary(volRep)
-	case v.instance.Spec.ReplicationState == volrep.Secondary:
+	case v.instance.Spec.ReplicationState == ramendrv1alpha1.Secondary:
 		return v.validateVRStatusAsSecondary(volRep)
-	case v.instance.Spec.ReplicationState == volrep.Resync:
-		return fmt.Errorf("resync is not supported for VRG")
 	default:
+		msg := "VolumeReplicationGroup state invalid"
+		v.updateProtectedPVCCondition(volRep.Name, PVCError, msg)
+
 		return fmt.Errorf("invalid Replication State %s for VolumeReplicationGroup (%s:%s)",
 			string(v.instance.Spec.ReplicationState), v.instance.Name, v.instance.Namespace)
 	}
 }
 
-// So, VolumeReplication controller reflects the status of a VolRep resource as primary
-// only after the promotion of the associated storage volume to primary is successful.
-// Hence, checking whether the primary state is reflected in the status is sufficient
-// to verify whether the VolRep resource is in a healthy state not. However, checking
-// conditions of VolRep status may be required for maintaining correct status of VRG
-// when VRG status is implemented.
-//
-// TODO: If checking conditions of VolRep is necessary for VRG status, add that logic
-//       here when VRG status is implemented.
+// VolumeReplication controller sets volRep.Status.State to primary only after the
+// successful promotion of the underlying storage volume consumed by the pvc.
+// It is necessary to get the appropriate condition out of different conditions
+// a VolRep status can have. Doing this may be helpful in getting a more
+// clear view of the VolumeReplication resource's current situation.
 func (v *VRGInstance) validateVRStatusAsPrimary(volRep *volrep.VolumeReplication) error {
```

Comment:
`validateVRStatusAsPrimar/Secondary` can be made into one function taking state as input, as rest of the code remains the same.

Reply:
ACK